dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTRequestContext.java

```diff
@@ -132,7 +133,31 @@ public void finish()
                     HttpSession s = (HttpSession)webappToSessionMap.get(wa);
                     IHttpSessionContext ctx =wa.getSessionContext();
                     if (ctx!=null){
-                    	wa.getSessionContext().sessionPostInvoke(s);
+                        // set app classloader as TCCL; JSF might need to load resources during session cleanup
+                        ClassLoader origClassLoader = ThreadContextHelper.getContextClassLoader();
+                        final ClassLoader warClassLoader = this.request.getServletContext().getClassLoader();
```

Comment:
Do you happen to know what type of loader is returned from `request.getServletContext().getClassLoader()`?  If it is an instance of `com.ibm.ws.classloading.internal.ThreadContextClassLoader`, then all is good.  If it is an instance of `com.ibm.ws.classloading.internal.AppClassLoader`, then I would suggest changing this line to something like:
```
ClassLoader appLoader = this.request.getServletContext().getClassLoader();
final ClassLoader warClassLoader = classLoadingService.createThreadContextClassLoader(appLoader); //despite the name, this will use the existing `ThreadContextClassLoader` if it is already created
```

The `ThreadContextClassLoader` can load additional classes beyond what the application classloader can.  It is possible that those extra classes may be required for de-serialization of session data.

The `ClassLoaderService` is a DS component that is used in `com.ibm.ws.webcontainer.osgi.WebContainer`.

Also, the call to `servletContext.getClassLoader()` will need to be in a doPriv block since getClassLoader is a privileged method call.

Reply:
Thanks for the thorough review.  I've just updated this `getClassLoader()` call - now it calls to the `WebApp` object `wa.getClassLoader()`, which has the app loader stored as a member.

In the _good_ case, both `origClassLoader` and `warClassLoader` end up being instances of
`com.ibm.ws.classloading.internal.ThreadContextClassLoader`.  Given this, I think it should be safe to assume that this `ThreadContextClassLoader` should always be returned from `wa.getClassLoader()` - what do you think?